From 608ba518b90b71e415bb9e0d553bd67415abd57b Mon Sep 17 00:00:00 2001
From: Lucian Zala <lucianx.zala@intel.com>
Date: Wed, 16 Sep 2015 11:57:08 +0300
Subject: [PATCH] Inform GPIO configuration via MDM_CTRL_GET_CFG

The current limitation is if  mcd is already initialized
user space will be wrongly informed about modem GPIO
configuration. This solution adds new ioctl command
MDM_CTRL_GET_CFG to handle GPIO configuration.

Change-Id: I86fd4f54e5d07660f94ed5d075adbbdd427dccb6
Signed-off-by: Lucian Zala <lucianx.zala@intel.com>
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-1646
---
 drivers/staging/modem_control/mdm_ctrl.c | 33 ++++++++++++++++++++++++++------
 drivers/staging/modem_control/mdm_util.h |  1 +
 include/linux/mdm_ctrl.h                 |  8 ++++++++
 3 files changed, 36 insertions(+), 6 deletions(-)

diff --git a/drivers/staging/modem_control/mdm_ctrl.c b/drivers/staging/modem_control/mdm_ctrl.c
index dfc8367783e3..1a86a7de8cea 100644
--- a/drivers/staging/modem_control/mdm_ctrl.c
+++ b/drivers/staging/modem_control/mdm_ctrl.c
@@ -293,6 +293,14 @@ static int get_hangup_reasons(struct mdm_info *mdm)
 	return mdm->hangup_causes;
 }
 
+/**
+ * mdm_ctrl_get_irq_info - get modem irq informations
+ */
+static unsigned char mdm_ctrl_get_irq_info(struct mdm_info *mdm)
+{
+	return mdm->irq_requests;
+}
+
 /*****************************************************************************
  *
  * Char device functions
@@ -362,6 +370,7 @@ inline bool mcd_is_initialized(struct mdm_info *mdm)
 static int mcd_init(struct mdm_info *mdm)
 {
 	int ret = 0;
+	mdm->irq_requests = NO_EVENT;
 
 	if (mdm->pdata->mdm.init(mdm->pdata->modem_data)) {
 		pr_err(DRVNAME ": MDM init failed...returning -ENODEV.");
@@ -392,11 +401,11 @@ static int mcd_init(struct mdm_info *mdm)
 				ret);
 			ret = -ENODEV;
 			goto del_cpu;
-		}
-	} else {
+		} else
+			mdm->irq_requests |= RST_EVENT;
+	} else
 		pr_info(DRVNAME ": No IRQ RST\n");
-		ret = -EIO;
-	}
+
 
 	if (mdm->pdata->cpu.get_irq_cdump(mdm->pdata->cpu_data) > 0) {
 		ret = request_irq(mdm->pdata->cpu.
@@ -410,7 +419,8 @@ static int mcd_init(struct mdm_info *mdm)
 				ret);
 			ret = -ENODEV;
 			goto free_irq;
-		}
+		} else
+			mdm->irq_requests |= CD_EVENT;
 	} else
 		pr_info(DRVNAME ": No IRQ COREDUMP\n");
 
@@ -502,7 +512,7 @@ long mdm_ctrl_dev_ioctl(struct file *filep, unsigned int cmd, unsigned long arg)
 
 			ret = mcd_init(mdm);
 
-			if ((!ret) || (ret == -EIO))
+			if (!ret)
 				pr_info(DRVNAME ": modem (board: %d, family: %d,pwr:%d,hub:%d)",
 					cfg.board, cfg.type,
 					cfg.pwr_on, cfg.usb_hub);
@@ -648,6 +658,17 @@ long mdm_ctrl_dev_ioctl(struct file *filep, unsigned int cmd, unsigned long arg)
 			mdm->polled_states);
 		break;
 
+	case MDM_CTRL_GET_CFG:
+		param = mdm_ctrl_get_irq_info(mdm);
+
+		ret = copy_to_user((void __user *)arg, &param, sizeof(param));
+		if (ret < 0) {
+			pr_info(DRVNAME ": copy to user failed ret = %ld\n",
+				ret);
+			return -EFAULT;
+		}
+		break;
+
 	case MDM_CTRL_SET_CFG:
 		pr_info(DRVNAME ": already configured\n");
 		ret = -EBUSY;
diff --git a/drivers/staging/modem_control/mdm_util.h b/drivers/staging/modem_control/mdm_util.h
index fadf3f26cb7c..f9e0162261f3 100644
--- a/drivers/staging/modem_control/mdm_util.h
+++ b/drivers/staging/modem_control/mdm_util.h
@@ -87,6 +87,7 @@ struct mdm_info {
 	struct timer_list flashing_timer;
 
 	bool is_mdm_ctrl_disabled;
+	unsigned char irq_requests;
 };
 
 /**
diff --git a/include/linux/mdm_ctrl.h b/include/linux/mdm_ctrl.h
index 5561da44f3bf..75271ef61932 100644
--- a/include/linux/mdm_ctrl.h
+++ b/include/linux/mdm_ctrl.h
@@ -77,6 +77,13 @@ enum mdm_ctrl_pwr_on_type {
 	POWER_ON_PMIC
 };
 
+/* Type of GPIO events configuration */
+enum {
+	NO_EVENT = 0x00,
+	RST_EVENT = 0x01,
+	CD_EVENT = 0x02,
+};
+
 /**
  * struct mdm_ctrl_cmd - Command parameters
  *
@@ -119,6 +126,7 @@ struct mdm_ctrl_cfg {
 #define MDM_CTRL_CLEAR_HANGUP_REASONS	_IO(MDM_CTRL_MAGIC, 9)
 #define MDM_CTRL_SET_POLLED_STATES	_IO(MDM_CTRL_MAGIC, 10)
 #define MDM_CTRL_SET_CFG		_IO(MDM_CTRL_MAGIC, 11)
+#define MDM_CTRL_GET_CFG		_IO(MDM_CTRL_MAGIC, 12)
 
 #endif /* _MDM_CTRL_H */
 
-- 
2.6.0.rc0.2.g7662973

